Develop stream 2024-06-26

Snektron opened the pull request:
@Naraenda is taking this over since @Snektron is on holiday.

This PR adds the following merge commits. **Emphasis** on important changes.

* 46faf89b **test: make tests pass with hardened libstc++** (resolves https://github.com/ROCm/rocPRIM/issues/570)
* 39c9d9bc ci(gitlab-ci.yml): re-enable most of the debug builds
* c6ea6aab test: improved test of device histogram_even for half/bfloat16
* a40cbdb9 fix: fix warnings when building rocprim from rocthrust when using hipcc
* 10c6175e **feat: add `partial_sort` and `partial_sort_copy` algorithms**
* 4a268483 **feat: add deterministic versions of `rocprim::inclusive_scan`, `rocprim::exclusive_scan`, `rocprim::inclusive_scan_by_key`, `rocprim::exclusive_scan_by_key`, and `rocprim::reduce_by_key`**
* 9c7c2970 style: format attributes on a separate line
* fc2de523 perf: remove oracle in nth element
* a4ae29cb ci(gitlab-ci.yml): compress autotuning artifacts before uploading them
* fab7fd63 refactor: improve code-reuse of `rocprim::detail::merge_path`
* f586d3ed test: add extra tests to nth-element
* d09bae6e **feat: add nth-element algorithm**
* 43c7ce71 perf: add auto tuning to reduce by key
* d63cd11a **feat: add large index support to device segmented reduce**
* 110f2d0f test: fix sizes generated by benchmark utility for random segments
* c523e13f ci: build everything using cmake-minimum and cmake-latest workflows
* d4aec9d4 test: fix benchmark minmax underoverflow
* 941984e4 test: disable naive batch memcpy in benchmarks
* 94e17b48 docs: document thread level methods
* f90e4e7e refactor: only autotune sleep-based lookback scan for gfx908
* 7b6b04bd docs: add documentation for autotuning
* e2aec7e6 test: improve test logging
* 1617f92a **refactor: deprecate raw_storage and replace it by uninitialized_array due to possible undefined behavior with raw_storage**
* f7209def **refactor: deprecate use of rocprim::thread_load and rocprim::thread_store**
* 917b1510 fix: fix block histogram test failures
* 441a934f docs: document thread types
* 2455a40e ci(gitlab-ci.yml): disable debug builds due to long compilation times
* 9fc5d182 fix: remove memory fence workaround on gfx10 and gfx11
* c07e212a fix: fix unroll warnings in debug min size releases
* ad746b46 **perf: add tuning to device partition**
* 9148c771 test: add benchmark flag to specify seed
* 9e849dbd docs: add contribution guidelines
* 71ba9cef **fix: prevent rocprim::adjacent_difference benchmarks from using more than 8 gb vram**
* d3b34418 test: fix windows test assertions
* c5a9d07c fix: disable force inline on windows
* 58c96d95 refactor: remove device device warpsize usage in host function in radix sort benchmark generator
* fc2ac733 **feat: add an overload of rocprim::match_any which accepts the maximum number of bits as a run-time parameter**
* 61ca0d13 **test: change default size in adjacent difference benchmark**
* 246fe37d fix: various fixes for batch memcpy
* de58f352 **perf: add auto tuning to transform**

Naraenda commented:
Rebased and added a fix that resolves https://github.com/ROCm/rocPRIM/issues/570

Naraenda commented:
Fixed doxygen errors in https://github.com/ROCm/rocPRIM/pull/575/commits/8d5b5e89021b08cb51d783cf0021294f545fac4c

Naraenda commented:
4651b07 Noticed CI failures after targeting a later commit in develop. `rocm-cmake` was recently-ish changed and now requires building. This commit resolves that when the `rocm-cmake` is freshly downloaded through `FetchContent`.

6edfe13 Should fix one of the test failures on MI200 with very large inputs.

Also rebased it, again.

CC @stanleytsang-amd

stanleytsang-amd commented:
> [4651b07](https://github.com/ROCm/rocPRIM/commit/4651b07a6af1642f0d0bc9d7166b4298d8eb4ade) Noticed CI failures after targeting a later commit in develop. `rocm-cmake` was recently-ish changed and now requires building. This commit resolves that when the `rocm-cmake` is freshly downloaded through `FetchContent`.
> 
> [6edfe13](https://github.com/ROCm/rocPRIM/commit/6edfe13e6c54fc564bf786200740410d962f36f4) Should fix one of the test failures on MI200 with very large inputs.
> 
> Also rebased it, again.
> 
> CC @stanleytsang-amd

Thanks Nara, I will rerun CI right now.

Naraenda commented:
I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 (4314c76). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.

CC @stanleytsang-amd

stanleytsang-amd commented:
> I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 ([4314c76](https://github.com/ROCm/rocPRIM/commit/4314c76ea790369a5b7819515aff5f5274c900aa)). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.
> 
> CC @stanleytsang-amd

Re-running CI now.  It does seem strange that an explicit hipStreamSynchronize is required here since this function only deals with one stream and all kernels are being executed on that stream and I don't see any data race conditions here.  Maybe we can make a small reproducer and file a bug report with the HIP team?

stanleytsang-amd commented:
> I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 ([4314c76](https://github.com/ROCm/rocPRIM/commit/4314c76ea790369a5b7819515aff5f5274c900aa)). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.
> 
> CC @stanleytsang-amd

It seems as though the hang still persists on MI200 on CI, even with the hipStreamSychronize().

Naraenda commented:
Rebased & resolved conflicts

stanleytsang-amd reviewed (changes requested):
Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.

Naraenda commented:
> Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.

Done in 8d06855.

stanleytsang-amd commented:
> > Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.
> 
> Done in [8d06855](https://github.com/ROCm/rocPRIM/commit/8d068554f5fe7cac67dda25d1f21723be7aadfd2).

Thanks.

stanleytsang-amd commented on benchmark/benchmark_config_dispatch.cpp:
```diff
@@ -91,20 +91,19 @@ int main(int argc, char** argv)
-
+        CREATE_BENCHMARK(async_stream, stream_kind::async_stream),
+        benchmark::RegisterBenchmark(
+            bench_naming::format_name("{lvl:na,algo:empty_kernel,cfg:default_config}"),
```
This does not compile:

/tmp/grpc/rocPRIM/benchmark/benchmark_config_dispatch.cpp:104:9: error: no matching function for call to 'RegisterBenchmark'
  104 |         benchmark::RegisterBenchmark(
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1182:29: note: candidate function not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1182 | inline internal::Benchmark* RegisterBenchmark(const char* name,
      |                             ^                 ~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1190:22: note: candidate function template not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1190 | internal::Benchmark* RegisterBenchmark(const char* name, Lambda&& fn) {
      |                      ^                 ~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1201:22: note: candidate function template not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1201 | internal::Benchmark* RegisterBenchmark(const char* name, Lambda&& fn,

Adding `.c_str()` fixes the compile error.

Beanavil replied:
Should be fixed now!

Naraenda committed:
feat(device_transform): add tuning benchmarks and config generation for device transform

Naraenda committed:
perf(device_transform): tuned device transform algorithm for better performance

Naraenda committed:
docs(changelog.md): add 'device_transform' improvements to changelog

Naraenda committed:
fix(transform_config_template): added missing '::type' in general case of transform config

Naraenda committed:
fix(benchmark_device_transform.cpp): fix unused type warning when compiling tuning benchmarks

Naraenda committed:
perf(device_transform.hpp): updated configs for device transform which uses a wider range of block sizes

Naraenda committed:
feat(ConfigAutotuneSettings.cmake): allow benchmark_device_transform to tune for more block sizes

Naraenda committed:
refactor(benchmark_device_transform.cpp): remove duplicated code with 'benchmark_device_transform.parallel.hpp'

Naraenda committed:
docs(changelog.md): removed 'slightly' in device transform performance improvements

Naraenda committed:
test(test_device_batch_memcpy.cpp): add simple batch copy test

This test can be more easily modified to find issues with batch (mem) copy.

Naraenda committed:
revert test(test_device_batch_memcpy.cpp): add simple batch copy test

This reverts commit 6dafd1c66684e775eae07fe4fd50632a80ca1673.

Naraenda committed:
fix(device_batch_memcpy.hpp): use dereference instead of 'rocprim::thread_load/store'

'thread_load/store' uses inline assembly prohibiting compiler optimization. This also bypasses an issue where 'thread_load' behaves oddly on debug builds.

Naraenda committed:
fix(benchmark_device_transform): fix various build errors and warnings

jblok27 committed:
Replaced section with match_any() call

jblok27 committed:
Added overload for match_any

Naraenda committed:
docs(changelog.md): update changelog with benchmark changes

Naraenda committed:
test(benchmark_device_adjacent_difference.cpp): increased the default size of input so that in place uint8 benchmarks don't fit in L3 cache on select architectures

jblok27 committed:
Fixed formatting

jblok27 committed:
change match_any to runtime dispatch

jblok27 committed:
Fixed copyright date

nolmoonen committed:
unified wavefront definition

Commit:
build: Remove force-inline workaround on windows

The problem mentioned there should be resolved by now.

Commit:
docs(intrinsics/warp): name the correct label_bits in match_any documentation

Commit:
docs: Add CHANGELOG for removing force-inline workaround

parbenc committed:
fix clang format

Commit:
fix(tests): Add saturating casts and use them for random data generation

The `static_casts` can over / underflow making the maximum value smaller
than the minimum. This was triggering an assert on the microsoft standard
library. Technically this was undefined behaviour that went unnoticed
on non-debug builds.

Saturate the input value to the range of the distribution type instead
to prevent this error.

Commit:
ci: enable debug builds on windows

Supposedly the slowest jobs should now be resolved, so this should work.

Beanavil committed:
Update contributing guidelines

Beanavil committed:
fix(benchmark_device_adjacent_difference): fixe size in bytes instead of number of elements

nolmoonen committed:
specify benchmark seed via command line

nolmoonen committed:
refactor lookback sleep dispatch

nolmoonen committed:
generic tuning

nolmoonen committed:
add config tuning for partition

nolmoonen committed:
add config to tests

ex-rzr committed:
Fix "warning: loop not unrolled" with CMAKE_BUILD_TYPE=MinSizeRel (-Os)

The compiler generates this warning when -Os is set:
warning: loop not unrolled: the optimizer was unable to perform the
requested transformation; the transformation might be disabled or
specified as part of an unsupported transformation ordering

Using static values as both loop bounds fixes these warnings. For some
reason, other optimization levels do not have this issue, the compiler
is able to understand that the number of iterations of the loops is
a compile-time value.

mfep committed:
fix(device_partition): re-added workaround for the device_partition family to properly limit block size for the base configuration

nolmoonen committed:
add tuned configurations

Naraenda committed:
ci(.gitlab-ciy.yml): disable debug builds in cmake-minimum due to excessive build times when targeting debug test

Naraenda committed:
refactor(intrinsics/thread.hpp): remove 'memory_fence_device' workaround for compiler bug on gfx10 and  gfx11

mfep committed:
Using .lint:clang-format

mfep committed:
fix(docs): Fixed documentation for thread subdir

NB4444 committed:
Deprecate thread_load/thread_store

mfep committed:
fix(docs): Fixed documentation for the types subdir

nolmoonen committed:
fix undefined behavior in test data generation

mfep committed:
Added ROCPRIM_DONT_SUPPRESS_DEPRECATIONS

mfep committed:
Deprecated raw_storage and replaced by uninitialized_array in a few locations

mfep committed:
unsigned char storage in raw_storage to prevent undefined aliasing

NB4444 committed:
Ignore thread_load and thread_store deprecation warnings

ex-rzr committed:
Refactor device_scan, use is_sleep_scan_state_used and with_scan_state as in other lookback algorithms

Commit:
Resolve "Improve rocPRIM test logs"

nolmoonen committed:
improve documentation for configuration tuning

Commit:
Resolve "Batch memcpy: disable BENCHMARK_BATCH_MEMCPY_NAIVE"

ex-rzr committed:
Do not build kernels with sleep in lookback state on devices that don't need it (!=gfx908)

ex-rzr committed:
Use device of the current stream in is_sleep_scan_state_used

Commit:
Resolve "Add thread headers to rocprim.hpp and document thread-level methods"

Commit:
Resolve "Benchmark utility for random segments generates segments of wrong size"

Commit:
Resolve "CMake build consistency"

Commit:
Resolve "Fix under- and overflow in minimum and maximum for input data for benchmarks"

Beanavil committed:
Update CHANGELOG

Beanavil committed:
Adapt device segmented_reduce for large indices within a segment

nolmoonen committed:
reduce by key tuning

Beanavil committed:
Add large indices test

NB4444 committed:
Tests nth element

NB4444 committed:
First commit nth element

NB4444 committed:
Simplified working version nth element on one block

NB4444 committed:
nth element sizes larger then 64

NB4444 committed:
Added equality buckets to nth_element logic

NB4444 committed:
Added output check for correctness

NB4444 committed:
Debugging synchronization

NB4444 committed:
Added multiple blocks for nth element

NB4444 committed:
Added test to see if elements did not change

NB4444 committed:
Nth element working version only for key with comperator greater and
less

NB4444 committed:
Nth element implemented for key with tests

NB4444 committed:
Fixed issue for custom types in nth element and added tests

NB4444 committed:
Added input and output itterators for nth element

NB4444 committed:
Added some benchmarks for nth element

NB4444 committed:
Made seperate kernel for block offset calculations nth element

NB4444 committed:
Small optimizations nth element

NB4444 committed:
Debug code nth element

NB4444 committed:
Small optimizations nth element

NB4444 committed:
Moved all block offset calculation to other kernel nth element

NB4444 committed:
Optimization nth element

NB4444 committed:
Make use of radix_rank instead of multiple scans

NB4444 committed:
Start of adding multiple items per thread nth element

NB4444 committed:
Nth element using less shared memory

NB4444 committed:
Nth element small optimizations and cleanup code

NB4444 committed:
Fixed benchmark break nth element after rebase

NB4444 committed:
nth element local oracle for buckets_store

NB4444 committed:
Cleanup nth element

NB4444 committed:
Nth element update tests with random nth element

NB4444 committed:
Add lookbackstates to nth element

NB4444 committed:
Cleanup and extra comments in nth_element

NB4444 committed:
Addition of configs for nth element

NB4444 committed:
Updated benchmark of nth_element based on feedback

NB4444 committed:
Added nth_element to changelog

NB4444 committed:
Nth_element updated tests and config based on review

NB4444 committed:
Removed unnecesarry test cases and choose nth_element based on seed_value

NB4444 committed:
Cleanup code nth element

NB4444 committed:
Nth element changes based on review

NB4444 committed:
Documentation updated for nth_element

NB4444 committed:
Changed nth element to a while loop

NB4444 committed:
Add documentation spinx doc

NB4444 committed:
Nth element asserts in device code

NB4444 committed:
Nth element documentation fixes

NB4444 committed:
nth element lookback state reset

NB4444 committed:
nth element docs crash fix

NB4444 committed:
Nth element changes based on review

NB4444 committed:
Replaced raw storage with unitialized_array in nth element

NB4444 committed:
Changed Nth element to be able to be used with iterators

NB4444 committed:
Added config for in place nth element

NB4444 committed:
Changes based on review

NB4444 committed:
nth element fix small mistakes

NB4444 committed:
Added c++17 tests nth_element

NB4444 committed:
Make use of internal merge_path also fix bug with unsigned types for size

NB4444 committed:
Added test for public merge_path_search

NB4444 committed:
Made review changes

NB4444 committed:
Fixed thread_load and thread_store bug with float and double

Snektron committed:
ci: compress autotune artifacts using zstd

NB4444 committed:
Add bug fixes to changelog

NB4444 committed:
Removed oracles array from nth element

Snektron committed:
ci: remove trailing newlines in gitlab-ci.yml

NB4444 committed:
Apply 1 suggestion(s) to 1 file(s)

Co-authored-by: Nara Prasetya <nara@streamhpc.com>

Snektron committed:
clang-format: trick clang-format into always breaking after c-style function attributes

NB4444 committed:
Remove constraint of 256 for number of buckets nth element

Snektron committed:
lookback scan: reformat

Makes formatting consistent with clang-format file.

Snektron committed:
lookback scan: remove HIP-CPU bits

- memcpy() (without std) works for HIP
- We don't really care about HIP-CPU anymore

This cleans up the source file a bit, and it doesn't seem like this
affects any benchmarks.

Snektron committed:
add ctz intrinsic

Counts the number of trailing zero bits. This is just
a clean wrapper around __builtin_ctz(ll).

Snektron committed:
test: print floats as hexfloat in assert_bit_eq

Snektron committed:
lookback scan: add reproducibility test

Snektron committed:
add warp_readfirstlane and warp_readlane intrinsics

Snektron committed:
lookback scan: add deterministic implementation

Snektron committed:
scan: add deterministic overload

Snektron committed:
scan_by_key: add deterministic overload

Snektron committed:
add char and short atomic load/store overloads

It seems that these are just supported and work fine.

Snektron committed:
reduce_by_key: add deterministic overload

Snektron committed:
lookback scan: swap flag and prefix, allow fast scan for values up to size 7

Since the prefix flag is always one byte now, we can put it behind the value
to get a smaller struct. This helps in some cases, for example, scan_by_key
over sizeof(AccType) = 2 now fits in an int instead of a long.

Snektron committed:
lookback scan: change flag to be always one byte

This slightly reduces the amount of memory required for a lookback
scan. Also, changing the INVALID value from -1 to 0xFF fixes some
sign issues there were before by using unsigned int as flag
underlying type.

Snektron committed:
nara nit f32

Snektron committed:
update changelog with mention of deterministic algorithms

Snektron committed:
lookback reproducibility test: allocate temporary memory with the right scan operator

Snektron committed:
lookback scan: avoid caching large types

These types are stored in a separate buffer, so we don't
need to or load them. Slightly speeds up deterministic
scan algorithms when the lookback scan type is > 7 bytes.

Snektron committed:
remove assertions in lookbacn scan, they don't compile properly in debug builds

Snektron committed:
lookback reproducibility test: use same functor for both tests

This enables the test to work with -ffast-math too.

Snektron committed:
lookback scan: rotate prefix rather than block_prefix

Snektron committed:
lookback scan: also test deterministic in normal tests

NB4444 committed:
partial sort benchmark

nolmoonen committed:
naive implementation

NB4444 committed:
Made partial_sort in place and created partial_sort_copy

NB4444 committed:
Test partial_sort with iterator

NB4444 committed:
Moved partial sort to own file

NB4444 committed:
Add partial_sort and partial_sort_copy to the changelog

NB4444 committed:
Add and fix documentation partial_sort

NB4444 committed:
Cleanup code based on nth_element review

NB4444 committed:
Merge with nth_element_remove_oracle branch

NB4444 committed:
Created c++17 test for partial_sort

NB4444 committed:
Added partial_sort_config

NB4444 committed:
Added benchmark for partial_sort

NB4444 committed:
Review adaptations

Melirius committed:
Restored tests for device histogram_even for half/bfloat16 types

parbenc committed:
add static_cast to fix compiler warning

NB4444 committed:
Fixed bug with inplicit casting in partial sort

NB4444 committed:
Removed unused variable and formatting

Naraenda committed:
test(test_device_batch_memcpy.cpp): standardize test names

Naraenda committed:
test(test_device_batch_memcpy.cpp): fix invalid calls being made to generate_random_data_n

ex-rzr committed:
ci: Enable debug builds excluding test_block_adjacent_difference/discontinuity

These tests take extremely long time to build with clang from ROCm 6.1+.

Naraenda committed:
test(test_intrinsics.cpp): fix invalid calls being made to test_utils::get_random_data

Naraenda committed:
docs: fix doxygen errors and warnings

Naraenda committed:
docs: update changelog

Naraenda committed:
ci(.gitlab-ci.yml): add hardened libc++ assertions when building tests with gitlab ci

Naraenda committed:
chore: bump version to 3.3.0

Naraenda committed:
build(cmake/Dependencies.cmake): build rocm-cmake depedency during populate step when fetching it by source

Naraenda committed:
refactor(benchmark_config_dispatch.cpp): fix unused variable and function

NB4444 committed:
Reduce items_per_thread for merge_sort to one for large types

NB4444 committed:
Reduce block_size for device_merge with large types
